### PR TITLE
[WIP]add url rewriting feature if APICAST_URL_REWRITE ENV is passed

### PR DESCRIPTION
**Pull request by kevprice83**
This feature is to allow URL rewriting when the same host value is used across multiple services and the path is used to route to the correct service and then the correct private backend.

### CODE REVIEW

**Comment by kevprice83** [comment]
@mikz this actually also includes a fix related to #254 I had to implement that for this not to fail

File: apicast/src/proxy.lua [hunk 1]
```diff
@@ -258,6 +261,18 @@ function _M.authorize(backend_version, service)
   end
 end
 
+local function rewrite_url(uri)
+  local i = find(uri, '/', 2)
+  
+  if i ~= nil and i > 0 then
```
Comment (mikz) [inline apicast/src/proxy.lua]: `i ~= nil` can be easily replaced by more readable `i`

File: apicast/src/proxy.lua [hunk 2]
```diff
@@ -258,6 +261,18 @@ function _M.authorize(backend_version, service)
...
+    path = sub(uri, i)
```
Comment (mikz) [inline apicast/src/proxy.lua]: path looks like global variable

**Comment by mikz** [comment]
Would be good to have some examples on how it rewrites.

**Comment by mayorova** [comment]
It will remove the first part of the path, for example:

`/first/second/third` => `/second/third`
`/single` => `/`
`/single-with-query?q=abc` => `?q=abc`

I think it's a very specific case, so it shouldn't be included to APIcast as global feature. But it would actually be good to have some kind of "plugin" that supports this feature, and that the user would be able to easily modify the  `rewrite_url` function.

**Comment by mikz** [comment]
I agree removing just the first part is too specific. 

We try to make people write less code so I'd like to figure out a way of defining this within concepts we already have - api upstream and mapping rules (including redirect_url).

Lets imagine an API:

```
/videos
/audio
/login
```

That is being exposed via a prefix:

```
/staging/video
/staging/audio
/staging/login
```

I guess the desire here would be to map this onto the existing api, stripping the `/staging/`.

Guess we could implement it in terms of `rewrite_url` of the mapping rule defined as `/staging/` with rewrite_url `/`. 

That way we could have arbitrary strings being replaced with other strings with no limits on where exactly that part is. And the replacement string could use values from the capture. Like `/staging/$api` => `/secret/$api`.

**Comment by kevprice83** [comment]
Closing this PR as it's too specific an implementation and we need something more generic.

**Pull request closed without merging**